distribution/lib/Standard/Base/0.0.0-dev/src/Error/Extensions.enso

```diff
@@ -156,3 +173,121 @@ Error.is_error = True
+   - handler: a callback which is called with the caught exception.
+
+   Returns the result of running `action` or the result of the `handler` if the
+   matching Java exception has been thrown from within the `action`.
+
+   > Example
+     Convert a string to an integer, catching `NumberFormatException` and
+     converting it to a more Enso-friendly dataflow error.
+
+     polyglot java import java.lang.Long
+     polyglot java import java.lang.NumberFormatException
+     parse str =
+         Panic.catch_java_exception NumberFormatException (Long.parseLong str) exception->
+             Error.throw (Illegal_Argument_Error "The provided string is not a valid number: "+exception.getMessage)
+Panic.catch_java_exception : Any -> Any -> (Any -> Any) -> Any
+Panic.catch_java_exception java_class ~action handler =
+    Panic.catch_any action caught_panic-> case caught_panic.payload of
+        Polyglot_Error java_exception ->
+            case Java.is_instance java_exception java_class of
+                True -> handler java_exception
+                False -> Panic.throw caught_panic
+        _ -> Panic.throw caught_panic
+
+## Executes the provided action and if a panic matching the provided type was
+   thrown, calls the provided callback.
+
+   If action executes successfully, the result of `Panic.catch` is the result of
+   that action. Otherwise, if a matching panic is thrown from within the action,
+   the result is obtained by calling the provided handler callback. Any
+   non-matching panics are forwarded without changes.
+
+   Arguments:
+   - typ: The expected panic type.
+   - action: The code to execute that potentially panics.
+   - handler: The callback to handle the panics. The callback will be provided
+     with a `Caught_Panic` instance encapsulating the `payload` of the caught
+     panic and its stacktrace.
+
+   > Example
+     Handling a specific type of panic.
+
+         Panic.catch Illegal_Argument_Error (Panic.throw (Illegal_Argument_Error "Oh no!" Nothing)) error->
+             IO.println "Caught an `Illegal_Argument_Error`: "+error.payload.message
+             "fallback value"
+Panic.catch : Any -> Any -> (Caught_Panic -> Any) -> Any
+Panic.catch typ ~action handler =
+    Panic.catch_any action caught_panic->
+        case caught_panic.payload.is_a typ of
+            True -> handler caught_panic
+            False -> Panic.throw caught_panic
+
+## Executes the provided action and converts any panic thrown by it into
+   an Error.
+
+   If action executes successfully, the result of `Panic.recover_any` is the
+   result of that action. Otherwise, it is the panic that was thrown after
+   conversion to a dataflow error.
+
+   Arguments:
+   - action: The code to execute that potentially panics.
+
+   > Example
+     Converting a panic to a dataflow error.
+
+         Panic.recover_any (Panic.throw "Oh no!")
+Panic.recover_any : Any -> Any
+Panic.recover_any ~action =
```

Comment:
Same with `catch_any`

Reply:
with `recover Any` - you are right, it's a leftover from when we have `Panic.recover` and I renamed it to `recover_any` and only then added `recover`. I definitely should remove `recover_any` and replace usages with `recover Any`. Will do.

As for catch - cannot do - `catch_any` is the primitve function _used to implement_ the selective `catch` which checks types. I can rename it to `primitive_catch` if you think that's better. 

Why I couldn't implement the typecheck in the primitive? Because `is_an` is implemented in-language and I didn't want two separate implementations of the same subtyping concept as I fear they may get out-of-sync. If we move `is_an` to the interpreter, `catch` itself could become a primitve removing the need for `catch_any`.